Build error when using `address` constraint with anchor V0.30.0

When using `anchor build` to compile the program,the fields limited by the `address constraint` cannot be recognized, but when compiled directly using `cargo build-bpf`, the compilation can be successful.
example:
```use anchor_lang::prelude::*;

declare_id!("2CuWx82vA1HdJ2tJQqd7E3LLRb4NsqknAGpJUX7ZUA7u");

#[program]
pub mod example {
    use super::*;

    pub fn initialize(_ctx: Context<Initialize>) -> Result<()> {
        Ok(())
    }
}

#[derive(Accounts)]
pub struct Initialize<'info> {
    pub owner: Signer<'info>,

    #[account(mut)]
    pub pool_state: Account<'info, PoolState>,

    /// CHECK:
    #[account(
        address = pool_state.token_vault
    )]
    pub vault: UncheckedAccount<'info>,
}

#[account()]
#[derive(Default, Debug)]
pub struct PoolState {
    pub token_vault: Pubkey,
}
```
Here is the error
```
error[E0425]: cannot find value `pool_state` in this scope
  --> programs/example/src/lib.rs:23:19
   |
23 |         address = pool_state.token_vault
   |                   ^^^^^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `example` (lib test) due to 1 previous error
Error: Building IDL failed
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,6 +50,7 @@ The minor version will be incremented upon a breaking change and the patch versi
 - cli: Fix upgradeable program clones ([#3010](https://github.com/coral-xyz/anchor/pull/3010)).
 - ts: Fix using IDLs that have defined types as generic arguments ([#3016](https://github.com/coral-xyz/anchor/pull/3016)).
 - idl: Fix generation with unsupported expressions ([#3033](https://github.com/coral-xyz/anchor/pull/3033)).
+- idl: Fix using `address` constraint with field expressions ([#3034](https://github.com/coral-xyz/anchor/pull/3034)).
 
 ### Breaking
 

--- a/lang/syn/src/idl/accounts.rs
+++ b/lang/syn/src/idl/accounts.rs
@@ -161,6 +161,7 @@ fn get_address(acc: &Field) -> TokenStream {
             .address
             .as_ref()
             .map(|constraint| &constraint.address)
+            .filter(|address| !matches!(address, syn::Expr::Field(_)))
             .map(|address| quote! { Some(#address.to_string()) })
             .unwrap_or_else(|| quote! { None }),
     }

--- a/tests/relations-derivation/programs/relations-derivation/src/lib.rs
+++ b/tests/relations-derivation/programs/relations-derivation/src/lib.rs
@@ -16,10 +16,16 @@ pub mod relations_derivation {
         ctx.accounts.account.bump = ctx.bumps.account;
         Ok(())
     }
+
     pub fn test_relation(_ctx: Context<TestRelation>) -> Result<()> {
         Ok(())
     }
+
     pub fn test_seed_constant(_ctx: Context<TestSeedConstant>) -> Result<()> {
+        Ok(())
+    }
+
+    pub fn test_address_relation(_ctx: Context<TestAddressRelation>) -> Result<()> {
         Ok(())
     }
 }
@@ -78,6 +84,14 @@ pub struct TestSeedConstant<'info> {
     )]
     account: Account<'info, MyAccount>,
     system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct TestAddressRelation<'info> {
+    #[account(address = my_account.my_account)]
+    account: UncheckedAccount<'info>,
+    #[account(seeds = [b"seed"], bump = my_account.bump)]
+    my_account: Account<'info, MyAccount>,
 }
 
 #[account]

--- a/tests/relations-derivation/tests/typescript.spec.ts
+++ b/tests/relations-derivation/tests/typescript.spec.ts
@@ -44,4 +44,9 @@ describe("typescript", () => {
   it("Can use relations derivation with seed constant", async () => {
     await program.methods.testSeedConstant().accounts({}).rpc();
   });
+
+  it("Can use relations derivation with `address` constraint", () => {
+    // Only compile test for now since the IDL spec doesn't currently support field access
+    // expressions for the `address` constraint
+  });
 });

